source/simulator/solver.cc

```diff
@@ -253,25 +253,91 @@ namespace aspect
       S_block_tolerance(S_block_tolerance)
     {}
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_A() const
     {
       return n_iterations_A_;
     }
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerA, class SInvOperator>
     unsigned int
-    BlockSchurPreconditioner<PreconditionerA, PreconditionerMp>::
+    BlockSchurPreconditioner<PreconditionerA, SInvOperator>::
     n_iterations_S() const
     {
-      return n_iterations_S_;
+      return S_inv_operator.n_iterations;
     }
 
-    template <class PreconditionerA, class PreconditionerMp>
+    template <class PreconditionerMp>
+
+    /* This class is used in the implementation of the right preconditioner.
+       the Schur complement is approximated by
+       the pressure mass matrix weighted by the inverse of viscosity.
+    */
+    class InverseWeightedMassMatrix
```

Comment:
This still isn't in the style we use to document classes. Please take a look at the style used in any of the header files.